### ISSUE TITLE
Support importing resource files bundled into Python packages

### ISSUE THREAD

**Issue by robinmackaij** [robotframework/robotframework#4317]
To better support distribution of "Resource bundles" between repos (e.g. to share Keywords between projects within the same company), enhancing the way `.resource` files are discovered and imported would be beneficial.

Building a package with the desired resource files is not very complex. A sample project structure (using `poetry`) could look like this:
```
src/
-- RobotFrameworkResources/
---- keywords/
------ __init__.py
------ Web.resource
---- variables/
------ __init__.py
------ Urls.resource
---- __init__.py
-- __init__.py
pyproject.toml
```
The part of the `pyproject.toml` for what to package would look like this (`[tool.poetry]` section):
``` toml
packages = [
    { include = "RobotFrameworkResources", from = "src" },
]
include = ["*.resource"]
```

A simple `poetry build` will generate the `.whl` and `.tar.gz` in the `dist` folder and can be distributed through regular means (`poetry publish` to upload to pypi or private tracker, adding them to a GitHub release, etc.).

Now where enhancement would be beneficial, is where the distributed resources would be used. Currently, getting access to the `Keywords` in the `Web.resource` would look like this:
``` robotframework
*** Settings ***
Resource    RobotFrameworkResources/keywords/Web.resource

*** Test Cases ***
Open the website
    Open main site

```
The disadvantage here is that the user of the resource package needs to know the internal structure of the distribution (`/keywords/`) and the extension of the resource file has to be specified.

An improved `Resource` import statement, more like the `Library` statement would be beneficial:
``` robotframework
*** Settings ***
Library  Collections
Library  RPA.Archive
Resource    Web
Resource    RobotFrameworkResource.Urls
```

Points for consideration:

- Which extensions to support? For initial implementation, I would choose `.resource` for simplicity / clarity.
- LSP / linter support
- How to allow package developers to "wire" the internal package (folder) structure to the desired Resource import statements

Edit:
Additional thoughts:

- How would package / library documentation be accessible to the user of the resource package? If the source package is not on GitHub or similar, any documentation distributed with the package (by, e.g. packaging `.md` or `.html` files) would be somewhere within a virtual environment / Python install folder of the user. How can we show this documentation?
    - perhaps the module docstring of the package can be shown by an LSP

@pekkaklarck From your Slack response, perhaps an idea is to follow the generic RF arguments syntax to allow users to specify which resources to import (no argument meaning "import all resources found in the package"):
``` robotframework
*** Settings ***
Resource    RobotFrameworkResource  # import all resources found in the resource package
```
``` robotframework
*** Settings ***
Resource    RobotFrameworkResource    Web.resource    Api.resource  # import only the specified resource files
```

**Comment by d-biehl** [comment]
I like the idea of handling resources the same way as Python packages, without the slash and without having to specify the file extension.
I am using "Robot Resource" packages in a current customer project. We also use Poetry to build the package and then install it into the venv where it is needed.

However, the slash and the file extension annoys me every time :-)

For wiring the packages, I would suggest something like __init__.py/robot. Maybe __init__.resource? And there should be something like __all__.

What's in the package and also showing the documentation of the package/resources should be part of the LSP implementation.

**Comment by pekkaklarck** [comment]
Being able to bundle resource files into Python packages would be useful. Packages are easy to install and resource file imports are convenient when there's no need to know where exactly they are located.

Implementing this ought to be easy by using [importlib.resources](https://docs.python.org/3/library/importlib.html#module-importlib.resources). Python's import machinery is also pretty fast so there shouldn't be performance problems. For example file system layout of the created package see:
https://importlib-resources.readthedocs.io/en/latest/using.html

I believe we should use the normal `Resource` setting so that it gets a single value similarly as with normal resource imports. This way imports look the same, overall syntax is compatible with earlier RF versions, and we can later come up with new usages for possible additional `Resource` parameters. We should, however, somehow explicitly separate the package name and the resource file name. I believe we could use a colon for that as in the following examples:

```
Resource        package:example.resource
Resource        package:another.resource
Resource        nested.package:example.rst
```

**Comment by pekkaklarck** [comment]
@robinmackaij and I discussed this at RoboCon sprints. When looking at alternatives, Robin pointed out that if a package like `package` is installed and it contains `example.resource` as package data, it can be already now imported like this:
```
Resource        package/example.resource
```
The reason this works is that we look for resource files from directories in the Python path and `package/example.resource` can be found there. It also works with nested structures like `nested/package/example.rst`. This syntax is so close to what I proposed above that it doesn't really make sense to add it. Let's just use what already works!

A drawback of the existing solution compared to what we could accomplish with the `package:example.resource` syntax and using [importlib.resources](https://docs.python.org/3/library/importlib.html#module-importlib.resources) is that the latter could work also if the package was a zip file. Unfortunately properly supporting that would be pretty hard, because our code currently expects there to be an actual file system path to the resource file. Even a bigger problem would be supporting the special `${CURDIR}` variable. This issue was about bundling resource files to Python packages in general and the benefits of the zip file support aren't worth the effort in that context. A separate issue can be submitted about that if someone is interested.

Interestingly it had never occurred to me that importing resource files bundled into Python packages already works like this and I'm sure there are others who aren't aware of it as well. I'll submit a separate issue about mentioning that in the User Guide. This issue can be closed after that.
